src/os/bluestore/BlueStore.cc

```diff
@@ -5672,6 +5687,112 @@ static void discard_cb(void *priv, void *priv2)
+      bufferlist header;
+      std::shared_lock l(c->lock);
+      OnodeRef onode = c->get_onode(*q, false);
+      {
+        KeyValueDB::Transaction txn = db->get_transaction();
+        uint64_t txn_cost = 0;
+        const string& prefix = Onode::calc_omap_prefix(onode->onode.flags);
+
+        KeyValueDB::Iterator it = db->get_iterator(prefix);
+        string head, tail;
+        onode->get_omap_header(&head);
+        onode->get_omap_tail(&tail);
+        it->lower_bound(head);
+        // head
+        if (it->valid() && it->key() == head) {
+          dout(30) << __func__ << "  got header" << dendl;
+          header = it->value();
+          if (header.length()) {
+            string new_head;
+            Onode::calc_omap_header(onode->onode.flags, onode.get(), &new_head, true);
+            if (new_head == head) {
+              continue;
+            }
+            txn->set(prefix, new_head, header);
+            txn_cost += new_head.length() + header.length();
+          }
+	        it->next();
+        }
+        // tail
+        {
+          string new_tail;
+          Onode::calc_omap_tail(onode->onode.flags, onode.get(), &new_tail, true);
+          bufferlist empty;
+          txn->set(prefix, new_tail, empty);
+          txn_cost += new_tail.length() + new_tail.length();
+        }
+        // values
+        string final_key;
+        Onode::calc_omap_key(onode->onode.flags, onode.get(), string(), &final_key, true);
+        size_t base_key_len = final_key.size();
+        while (it->valid() && it->key() < tail) {
+          string user_key;
+          onode->decode_omap_key(it->key(), &user_key);
+          dout(20) << __func__ << "  got " << pretty_binary_string(it->key())
+            << " -> " << user_key << dendl;
+
+          final_key.resize(base_key_len);
+          final_key += user_key;
+          auto v = it->value();
+          txn->set(prefix, final_key, v);
+          txn_cost += final_key.length() + v.length();
+
+          // submit a portion if cost exceeds 16MB
+          if (txn_cost >= 16 * (1 << 20) ) {
+            db->submit_transaction_sync(txn);
+            txn = db->get_transaction();
+            txn_cost = 0;
+          }
+          it->next();
+        }
+        if (txn_cost > 0) {
+	        db->submit_transaction_sync(txn);
+        }
+      }
+      // finalize: remove legacy data
+      {
+        KeyValueDB::Transaction txn = db->get_transaction();
+        // remove old keys
+        const string& prefix = onode->get_omap_prefix();
+        string old_head, old_tail;
+        onode->get_omap_header(&old_head);
+        onode->get_omap_tail(&old_tail);
+        txn->rm_range_keys(prefix, old_head, old_tail);
+        txn->rmkey(prefix, old_tail);
+        _record_onode(onode, txn);
+        db->submit_transaction_sync(txn);
+      }
+    }
+  }
+  // when fix finish, we should change omap_legacy
+  omap_legacy = false;
+  write_meta("omap_type", "new");
```

Comment:
previously we kept store wide omap format type within a superblock persisted in RocksDB under PREFIX_SUPER. Suggest to follow the same procedure.